dev/DynamicDependency/DataStore.cpp

```diff
@@ -0,0 +1,167 @@
+
+#include "pch.h"
+
+#include "DataStore.h"
+
+#include "DynamicDependencyDataStore_h.h"
+#include "utf8.h"
+#include "winrt_msixdynamicdepednency.h"
+
+#include <wil/winrt.h>
+
+#include <shlobj.h>
+
+MddCore::PackageDependency MddCore::DataStore::Load(PCWSTR packageDependencyId)
+{
+    std::filesystem::path relativeFilename{ L"DynamicDependency" };
+    relativeFilename /= std::wstring(packageDependencyId) + DataStore::fileExtension;
+
+    auto filename{ GetDataStorePathForUser() };
+    filename /= relativeFilename;
+    wil::unique_hfile file{ OpenFileIfExists(filename.c_str()) };
+    if (!file)
+    {
+        filename = GetDataStorePathForSystem();
+        filename /= relativeFilename;
+        file.reset(OpenFileIfExists(filename.c_str()));
+        if (!file)
+        {
+            // Not found
+            return PackageDependency();
+        }
+    }
+
+    LARGE_INTEGER fileSize{};
+    THROW_IF_WIN32_BOOL_FALSE(::GetFileSizeEx(file.get(), &fileSize));
+    const auto dataSize{ fileSize.QuadPart };
+    THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_DATA), dataSize > INT32_MAX);
+    if (dataSize == 0)
+    {
+        // 0-byte file is invalid. Perhaps power was lost when written but before flushed?
+        // 'Fix' it i.e. delete it and report not-found
+        file.reset();
+        std::filesystem::remove(filename);
+        return PackageDependency();
+    }
+
+    const auto bufferSize{ static_cast<DWORD>(dataSize) + 1 };
+    std::unique_ptr<char[]> bufferUtf8{ new char[bufferSize] };
+
+    DWORD bytesRead{};
+    THROW_IF_WIN32_BOOL_FALSE(::ReadFile(file.get(), bufferUtf8.get(), bufferSize, &bytesRead, nullptr));
+    file.reset();
+    bufferUtf8[bytesRead] = '\0';
+    auto json{ bufferUtf8.get() };
+
+    return MddCore::PackageDependency::FromJSON(json);
+}
+
+void MddCore::DataStore::Save(
+    const MddCore::PackageDependency& packageDependency,
+    const MddCreatePackageDependencyOptions options)
+{
+    // Package depednencies with LifetimeKind.Process are never saved to disk. They live in memory and end with the process
+    if (packageDependency.LifetimeKind() == MddPackageDependencyLifetimeKind::Process)
+    {
+        return;
+    }
+
+    auto json{ packageDependency.ToJSONUtf8() };
+
+    auto path{ GetDataStorePath(options) };
+    path /= L"DynamicDependency";
+    std::filesystem::create_directory(path);
+
+    auto filename{ path / (packageDependency.Id() + DataStore::fileExtension) };
+
+    wil::unique_hfile file{ ::CreateFileW(filename.c_str(), GENERIC_WRITE, FILE_SHARE_DELETE, nullptr, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, nullptr) };
+    if (!file)
+    {
+        THROW_HR_MSG(HRESULT_FROM_WIN32(GetLastError()), "%ls", filename.c_str());
+    }
+
+    DWORD bytesWritten{};
+    THROW_IF_WIN32_BOOL_FALSE_MSG(::WriteFile(file.get(), json.c_str(), static_cast<DWORD>(json.length()), &bytesWritten, nullptr), "%ls", filename.c_str());
+}
+
+void MddCore::DataStore::Delete(PCWSTR packageDependencyId)
```

Comment:
nit: Prefer passing const& wstring_view instead of PCWSTR (applies to this whole PR). It works with all the things that need PCWSTR and it's safer (and supports non-null-terminated/sliced strings which can be nice).

Reply:
Will consider in future PR